src/DoctrineBatchUtils/BatchProcessing/SelectBatchIteratorAggregate.php

```diff
@@ -0,0 +1,138 @@
+<?php
+declare(strict_types=1);
+
+namespace DoctrineBatchUtils\BatchProcessing;
+
+use ArrayIterator;
+use DoctrineBatchUtils\BatchProcessing\Exception\MissingBatchItemException;
+use Doctrine\ORM\AbstractQuery;
+use Doctrine\ORM\EntityManagerInterface;
+use IteratorAggregate;
+use Traversable;
+
+/**
+ * 'Read' focused batching iterator that will issue a clear to the entity manager
+ * every batch size to 'detach' the managed objects and therefore make them entities only for the
+ * purpose of reading without manually attaching them back to the EntityManager.
+ */
+final class SelectBatchIteratorAggregate implements IteratorAggregate
+{
+    /**
+     * @var Traversable
+     */
+    private $resultSet;
+
+    /**
+     * @var EntityManagerInterface
+     */
+    private $entityManager;
+
+    /**
+     * @var int
+     */
+    private $batchSize;
+
+    public static function fromQuery(AbstractQuery $query, int $batchSize): self
+    {
+        return new self($query->iterate(), $query->getEntityManager(), $batchSize);
+    }
+
+    /**
+     * @param object[] $results
+     * @param EntityManagerInterface $entityManager
+     * @param int $batchSize
+     *
+     * @return self
+     */
+    public static function fromArrayResult(
+        array $results,
+        EntityManagerInterface $entityManager,
+        int $batchSize
+    ): self {
+        return new self(new ArrayIterator($results), $entityManager, $batchSize);
+    }
+
+    public static function fromTraversableResult(
+        Traversable $results,
+        EntityManagerInterface $entityManager,
+        int $batchSize
+    ): self {
+        return new self($results, $entityManager, $batchSize);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getIterator()
```

Comment:
Return type

Reply:
The upstream interface is not typed so it wouldn't be fitting for me to decide the real type. If you want I can though and go with `Traversable`

Reply:
You can use `: iterable` as return type